stackslib/src/chainstate/nakamoto/staging_blocks.rs

```diff
@@ -463,6 +509,49 @@ impl<'a> NakamotoStagingBlocksConnRef<'a> {
         let res: Option<u64> = query_row(self, sql, args)?;
         Ok(res.is_some())
     }
+
+    /// Is this a shadow tenure?
+    /// If any block is a shadow block in the tenure, they must all be.
+    ///
+    /// Returns true if the tenure has at least one shadow block.
+    pub fn is_shadow_tenure(
+        &self,
+        consensus_hash: &ConsensusHash,
+    ) -> Result<bool, ChainstateError> {
+        let qry = "SELECT 1 FROM nakamoto_staging_blocks WHERE consensus_hash = ?1 AND obtain_method = ?2";
+        let args = rusqlite::params![
+            consensus_hash,
+            NakamotoBlockObtainMethod::Shadow.to_string()
+        ];
+        let present: Option<u32> = query_row(self, qry, args)?;
+        Ok(present.is_some())
+    }
+
+    /// Shadow blocks, unlike Stacks blocks, have a unique place in the chain history.
+    /// They are inserted post-hoc, so they and their underlying burnchain blocks don't get
+    /// invalidated via a fork.  A consensus hash can identify (1) no tenures, (2) a single
+    /// shadow tenure, or (3) one or more non-shadow tenures.
+    ///
+    /// This is important when downloading a tenure that is ended by a shadow block, since it won't
+    /// be processed beforehand and its hash isn't learned from the burnchain (so we must be able
+    /// to infer that if this is a shadow tenure, none of the blocks in it have siblings).
+    pub fn get_shadow_tenure_start_block(
```

Comment:
Is it possible to have some unit tests for the database calls?

Not sure how easy that is to set up, but would be nice.